letsencrypt/models/letsencrypt.py

```diff
@@ -0,0 +1,157 @@
+        account_key = os.path.join(data_dir, 'account.key')
+        if not os.path.isfile(account_key):
+            _logger.info('generating rsa account key')
+            self.call_cmdline([
+                'openssl', 'genrsa', '-out', account_key,
+                str(DEFAULT_KEY_LENGTH),
+            ])
+        assert os.path.isfile(account_key), 'failed to create rsa key'
+        return account_key
+
+    @api.model
+    def generate_domain_key(self, domain):
+        domain_key = os.path.join(self.get_data_dir(), '%s.key' % domain)
+        if not os.path.isfile(domain_key):
+            _logger.info('generating rsa domain key for %s', domain)
+            self.call_cmdline([
+                'openssl', 'genrsa', '-out', domain_key,
+                str(DEFAULT_KEY_LENGTH),
+            ])
+        return domain_key
+
+    @api.model
+    def validate_domain(self, domain):
+        if domain in ['localhost', '127.0.0.1'] or\
+                domain.startswith('10.') or\
+                domain.startswith('172.16.') or\
+                domain.startswith('192.168.'):
+            raise exceptions.Warning(
+                _("Let's encrypt doesn't work with private addresses!"))
+
+    @api.model
+    def generate_csr(self, domain):
+        domains = [domain]
+        i = 0
+        while self.env['ir.config_parameter'].get_param(
+                'letsencrypt.altname.%d' % i):
+            domains.append(
+                self.env['ir.config_parameter']
+                .get_param('letsencrypt.altname.%d' % i)
+            )
+            i += 1
+        _logger.info('generating csr for %s', domain)
+        if len(domains) > 1:
+            _logger.info('with alternative subjects %s', ','.join(domains[1:]))
+        config = self.env['ir.config_parameter'].get_param(
+            'letsencrypt.openssl.cnf', '/etc/ssl/openssl.cnf')
+        csr = os.path.join(self.get_data_dir(), '%s.csr' % domain)
+        with tempfile.NamedTemporaryFile() as cfg:
+            cfg.write(open(config).read())
+            if len(domains) > 1:
+                cfg.write(
+                    '\n[SAN]\nsubjectAltName=' +
+                    ','.join(map(lambda x: 'DNS:%s' % x, domains)) + '\n')
+            cfg.file.flush()
+            cmdline = [
+                'openssl', 'req', '-new',
+                self.env['ir.config_parameter'].get_param(
+                    'letsencrypt.openssl.digest', '-sha256'),
+                '-key', self.generate_domain_key(domain),
+                '-subj', '/CN=%s' % domain, '-config', cfg.name,
+                '-out', csr,
+            ]
+            if len(domains) > 1:
+                cmdline.extend([
+                    '-reqexts', 'SAN',
+                ])
+            self.call_cmdline(cmdline)
+        return csr
+
+    @api.model
+    def cron(self):
+        domain = urlparse.urlparse(
+            self.env['ir.config_parameter'].get_param(
+                'web.base.url', 'localhost')).netloc
```

Comment:
letsencrypt simply doesn't support port numbers in host names. You need a server listening on port 80. This should be added to https://github.com/OCA/server-tools/pull/347/files#diff-9b7e43b14857ae1b9294fe0125a6846fR73 I think